libcrux-psq/README.md

```diff
@@ -2,10 +2,235 @@
-attackers.
+This crate implements a protocol for establishing, and registering a
+post-quantum shared secret between an initiator and a responder.
+
+## Query Mode
+
+The outcome of a query run is that the initiator can send one payload
+to the responder, which can return one response to the initiator.
+
+These payloads **do not** enjoy post quantum protection.
+
+```
+Common inputs:
+    - ctx
+
+Inputs of A:
+    - query_payload
+    - query_aad
+    - pk_B = g^s
+    
+Inputs of B: 
+    - (sk_B = s, pk_B = g^s)
+    - response_aad
+    - query handler f
+
+A: 
+    (esk_A = x, epk_A = g^x) <- DH.KeyGen()
+    tx0 = hash(0 | ctx | pk_B | epk_A)
+    dh_shared_secret_query = DH.Derive(esk_A, pk_B)
+    K_0  = KDF(dh_shared_secret_query, tx0)
+    (enc_query, tag_query) <- AEAD.Encrypt(K_0, query_payload, query_aad)
+    
+A -> B: (epk_A, enc_query, tag_query, query_aad)
+
+B:
+    tx0 = hash(0 | ctx | pk_B | epk_A)
+    dh_shared_secret_query = DH.Derive(sk_B, epk_A)
+    K_0 = KDF(dh_shared_secret_query)
+    query_payload = AEAD.Decrypt(K_0, enc_query, tag_query, query_aad)
+    ...
+    response_payload <- f(query_payload)
+    ...
+    (esk_B = y, epk_B = g^y) <- DH.KeyGen()
+    tx2 = hash(2 | tx0 | epk_B)
+    dh_shared_secret_response_1 = DH.Derive(sk_B, epk_A)
+    dh_shared_secret_response_2 = DH.Derive(esk_B, epk_A)
+    K_2 = KDF(K_0 | dh_shared_secret_response_1 | dh_shared_secret_response_2, tx2)
+    (enc_response, tag_response) <- AEAD.Encrypt(K_2, response_payload, response_aad)
+
+B -> A: (epk_B, enc_response, tag_response, response_aad)
+
+A:
+    tx2 = hash(2 | tx0 | epk_B)
+    dh_shared_secret_response_1 = DH.Derive(esk_A, pk_B)
+    dh_shared_secret_response_2 = DH.Derive(esk_A, epk_B)
+    K_2 = KDF(K_0 | dh_shared_secret_response_1 | dh_shared_secret_response_2, tx2)
+    response_payload = AEAD.Decrypt(K_2, enc_response, tag_response, response_aad)
+```
+
+## Registration Mode
+
+The outcome of a query run is that initiator and responder agree on a
+shared secret `K_2` and the initiator can additionally send one
+payload to the responder, which can return one response to the
+initiator.
+
+Both payloads enjoy HNDL protection, if a PQ-secure inner KEM is
+used.
+
+The shared secret can be used to derive a large number of secure
+transport sessions between initiator and responder (see below).
+
+### Handshake
+```
+Common Inputs:
+    - ctx
+    
+Inputs of A: 
+    - registration_payload
+    - registration_outer_aad
+    - registration_inner_aad
+    - pk_B = g^s
+    - (sk_A = c, pk_A = g^c)
+    - pqpk_B (optional)
+    
+Inputs of B: 
+    - (sk_B = s, pk_B = g^s)
+    - (pqsk_B, pqpk_B)
+    - query handler f,
+    - response_aad, 
+    - registration handler f
+    
+A: 
+    (esk_A = x, epk_A = g^x) <- DH.KeyGen()
+    tx0 = hash(0 | ctx | pk_B | epk_A)
+    dh_shared_secret_outer = DH.Derive(esk_A, pk_B)
+    K_0  = KDF(dh_shared_secret_outer, tx0)
+    if pqpk_S provided
```

Comment:
This is not an input. So where is it coming from?
Making indices for parties a little clearer overall, also who is playing which role, would help in general as well.